hashjoin.go

```diff
@@ -0,0 +1,393 @@
+
+	for i := 0; i < hashTable.nOutCols; i++ {
+		hashTable.values[i] = append(hashTable.values[i], make(intColumn, amount)...)
+	}
+
+	hashTable.allocated += amount
+}
+
+func (hashTable *hashTableInt) growNext(amount int) {
+	hashTable.next = append(hashTable.next, make([]int, amount)...)
+}
+
+func (hashTable *hashTableInt) insertKey(hashKey int, id int) {
+	hashTable.count++
+	hashTable.next[id] = hashTable.first[hashKey]
+	hashTable.first[hashKey] = id
+}
+
+func makeHashJoinGroupBuilder(hashTable *hashTableInt) *hashJoinGroupBuilder {
+	return &hashJoinGroupBuilder{
+		hashTable: hashTable,
+	}
+}
+
+func (builder *hashJoinGroupBuilder) insertBatch(flow dataFlow, eqColIdx int, outCols []int, batchSize int) {
+	builder.hashTable.growGroup(batchSize)
+
+	eqCol := flow.b[eqColIdx].(intColumn)[:batchRowLen]
+
+	for i := 0; i < batchSize; i++ {
+		builder.hashTable.keys[i + builder.totalSize + 1] = eqCol[i]
+	}
+
+	for idx, colIdx := range outCols {
+		outCol := flow.b[colIdx].(intColumn)[:batchRowLen]
+		builder.spread(outCol, idx, batchSize)
+	}
+
+	builder.totalSize += batchSize
+}
+
+func (builder *hashJoinGroupBuilder) spread(outCol intColumn, valIdx int, batchSize int) {
+	valCol := builder.hashTable.values[valIdx]
+
+	for i:= 0; i < batchSize; i++ {
+		valCol[i + builder.totalSize + 1] = outCol[i]
+	}
+}
+
+func (builder *hashJoinGroupBuilder) computeBucket() {
+	builder.bucket = make([]int, builder.totalSize)
+
+	for i := 0; i < builder.totalSize; i++ {
+		builder.bucket[i] = builder.hashTable.hashInt(builder.hashTable.keys[i + 1])
+	}
+}
+
+func (builder *hashJoinGroupBuilder) insertKeys() {
+	builder.hashTable.growNext(builder.totalSize)
+
+	for i := 0; i < builder.totalSize; i++ {
+		builder.hashTable.insertKey(builder.bucket[i], i + 1)
+	}
+}
+
+func (builder *hashJoinGroupBuilder) start(source ExecOp, eqColIdx int, outCols []int) {
+	for{
+		flow := source.Next()
+		batchSize := flow.n
+
+		if batchSize == 0 {
+			break
+		}
+
+		builder.insertBatch(flow, eqColIdx, outCols, batchSize)
+	}
+
+	builder.computeBucket()
+	builder.insertKeys()
+}
+
+// hashJoinerIntInt performs a hashJoin joining on two integer columns where the
+// left table represents the build relation. It does not work with N - N joins.
+type hashJoinerIntInt struct {
+	hashTable *hashTableInt
+
+	left ExecOp
+	right ExecOp
+
+	leftEqColIdx  int
+	rightEqColIdx int
+
+	leftCols []int
+	rightCols []int
+
+	outFlow dataFlow
+}
+
+func (hashJoiner *hashJoinerIntInt) Init() {
+	nOutputCols := len(hashJoiner.leftCols) + len(hashJoiner.rightCols)
+	if nOutputCols == 0 {
+		panic("no output cols")
+	}
+
+	hashJoiner.outFlow.b = make(batch, nOutputCols)
+	for i := range hashJoiner.outFlow.b {
+		hashJoiner.outFlow.b[i] = make(intColumn, batchRowLen)
+	}
+
+	// todo(changangela): we currently assume build relation has unique join
+	// column and is on the left
+
+	hashJoiner.hashTable = makeHashTableInt(hashTableBucketSize, len(hashJoiner.leftCols))
+
+	hashJoiner.build()
```

Comment:
This doesn't belong in `Init`, which is designed to run before execution starts at all - more of a setup phase than a do work phase. You should put this into `Next` behind a conditional that will only run once. In distsql we do this with a little state machine infrastructure.